src/trusted/service_runtime/nacl_syscall_common.c

```diff
@@ -578,67 +641,100 @@ int32_t NaClSysOpen(struct NaClAppThread  *natp,
-      NaClLog(1, "Entered directory into open file table at %d\n",
-              retval);
+    if (!retval) {
+      retval = NaClSetAvail(nap, ((struct NaClDesc *) NaClDescDirDescMake(hd)));
+      NaClLog(1, "Entered directory into open file table at %d\n", retval);
     }
   } else {
     struct NaClHostDesc  *hd;
 
     hd = malloc(sizeof *hd);
-    if (NULL == hd) {
+    if (!hd) {
       retval = -NACL_ABI_ENOMEM;
       goto cleanup;
     }
     retval = NaClHostDescOpen(hd, path, flags, mode);
-    NaClLog(1,
-            "NaClHostDescOpen(0x%08"NACL_PRIxPTR", %s, 0%o, 0%o) returned %d\n",
+    NaClLog(1, "NaClHostDescOpen(0x%08"NACL_PRIxPTR", %s, 0%o, 0%o) returned %d\n",
             (uintptr_t) hd, path, flags, mode, retval);
-    if (0 == retval) {
-      retval = NaClSetAvail(nap,
-                            ((struct NaClDesc *) NaClDescIoDescMake(hd)));
+    if (!retval) {
+      retval = NaClSetAvail(nap, ((struct NaClDesc *) NaClDescIoDescMake(hd)));
       NaClLog(1, "Entered into open file table at %d\n", retval);
     }
   }
+
 cleanup:
-  return retval;
+  // yiwen: now translate the real fds to virtual fds and return them to the cage
+  fd_cage_table[nap->cage_id][nap->fd] = retval;
+  fd_retval = nap->fd;
+  nap->fd++;
+
+  // yiwen: register the fd and lib_path info for the cage, in lib_table[CACHED_LIB_NUM_MAX]
+  //        this will be used when trying to check if a lib has been cached in our system
+  // yiwen: do sanity check for the given fd first before our registration
+  if (fd_retval > CACHED_LIB_NUM_MAX) {
+     strncpy(nap->lib_table[fd_retval].path, path, strlen(path) + 1);
+     nap->num_lib++;
+  }
+
+  NaClLog(1, "[*** Debug ***][Open] fd = %d, filepath = %s \n", fd_retval, path);
+
+  return fd_retval;
 }
 
-int32_t NaClSysClose(struct NaClAppThread *natp,
-                     int                  d) {
+int32_t NaClSysClose(struct NaClAppThread *natp, int d) {
   struct NaClApp  *nap = natp->nap;
+  struct NaClDesc *ndp = NULL;
   int             retval = -NACL_ABI_EBADF;
-  struct NaClDesc *ndp;
+  int             fd = 0;
 
-  NaClLog(3, "Entered NaClSysClose(0x%08"NACL_PRIxPTR", %d)\n",
+  NaClLog(1, "Entered NaClSysClose(0x%08"NACL_PRIxPTR", %d)\n",
           (uintptr_t) natp, d);
 
   NaClFastMutexLock(&nap->desc_mu);
-  ndp = NaClGetDescMu(nap, d);
-  if (NULL != ndp) {
-    NaClSetDescMu(nap, d, NULL);  /* Unref the desc_tbl */
+
+  /* debug -jp */
+  if (nap->cage_id) {
+     NaClLog(1, "cage_id: %d\n", nap->cage_id);
+     retval = 0;
+     goto out;
   }
-  NaClFastMutexUnlock(&nap->desc_mu);
-  NaClLog(5, "Invoking Close virtual function of object 0x%08"NACL_PRIxPTR"\n",
-          (uintptr_t) ndp);
-  if (NULL != ndp) {
+
+  // yiwen
+  fd = fd_cage_table[nap->cage_id][d];
+
+  /* debug -jp */
+  if (fd == 8000 || fd == 8001) {
+     NaClLog(1, "cage_id: %d fd = %d\n", nap->cage_id, fd);
+     retval = 0;
+     goto out;
+  }
```

Comment:
I would mark this block with a FIXME or a TODO in the comment.

Reply:
Will do